File: mobile/lib/src/core/phone/cambodian_phone.dart
/// Cambodian mobile numbers, as the Telecommunication Regulator of Cambodia allocates them:
/// https://www.trc.gov.kh/en/resources/mobile-prefixes/ (read 2026-09-25).
///
/// A number is a three-digit prefix (with its leading 0) and a subscriber part of six digits,
/// or seven for the prefixes TRC marks with an asterisk. Anything else is not a Cambodian
/// mobile, and the request form is the one place a wrong digit costs a life: the donor who
/// accepted calls it and nobody answers. Mirrors `CambodianPhone.java` on the server — change
/// both together.
abstract final class CambodianPhone {
    static const Set<String> _sixDigit = {
        '010', '011', '012', '014', '015', '016', '017', '060', '061', '066', '067', '068',
        '069', '070', '077', '078', '081', '085', '086', '087', '089', '090', '092', '093',
        '095', '098', '099',
    };
    static const Set<String> _sevenDigit = {'018', '031', '071', '076', '088', '096', '097'};

    /// The number in `+855XXXXXXXX` form, or null if [input] is not a valid Cambodian mobile.
    ///
    /// Accepts what people actually type: `012 345 678`, `012-345-678`, `+855 12 345 678`,
    /// `855012345678`, `00855 12345678`.
    static String? normalize(String input) {
        var digits = input.trim();
        if (!RegExp(r'^\+?[0-9 .\-()]+$').hasMatch(digits)) {
            return null;
        }
        digits = digits.replaceAll(RegExp(r'[^0-9]'), '');
        if (digits.startsWith('00855')) {
            digits = digits.substring(5);
        } else if (digits.startsWith('855')) {
            digits = digits.substring(3);
        }
        if (!digits.startsWith('0')) {
            digits = '0$digits';
        }
        if (digits.length < 3) {
            return null;
        }
        final prefix = digits.substring(0, 3);
        final subscriberLength = digits.length - 3;
        final valid = (_sixDigit.contains(prefix) && subscriberLength == 6) ||
            (_sevenDigit.contains(prefix) && subscriberLength == 7);
        return valid ? '+855${digits.substring(1)}' : null;
    }

    static bool isValid(String input) => normalize(input) != null;
}

File: backend/src/main/java/kh/lifelink/api/request/CambodianPhone.java
package kh.lifelink.api.request;

import java.util.Optional;
import java.util.Set;
import java.util.regex.Pattern;

/**
 * Cambodian mobile numbers, as the Telecommunication Regulator of Cambodia allocates them:
 * https://www.trc.gov.kh/en/resources/mobile-prefixes/ (read 2026-09-25).
 *
 * <p>A number is a three-digit prefix (with its leading 0) and a subscriber part of six digits, or
 * seven for the prefixes TRC marks with an asterisk. The request's callback number is the one the
 * accepting donor dials, so a wrong one is worse than none. Mirrors {@code cambodian_phone.dart} in
 * the mobile app — change both together.
 */
final class CambodianPhone {

    private static final Set<String> SIX_DIGIT =
            Set.of(
                    "010", "011", "012", "014", "015", "016", "017", "060", "061", "066", "067",
                    "068", "069", "070", "077", "078", "081", "085", "086", "087", "089", "090",
                    "092", "093", "095", "098", "099");
    private static final Set<String> SEVEN_DIGIT =
            Set.of("018", "031", "071", "076", "088", "096", "097");
    private static final Pattern TYPED = Pattern.compile("^\\+?[0-9 .\\-()]+$");

    private CambodianPhone() {}

    /** The number as {@code +855XXXXXXXX}, or empty if it is not a Cambodian mobile. */
    static Optional<String> normalize(String input) {
        if (input == null || !TYPED.matcher(input.trim()).matches()) {
            return Optional.empty();
        }
        String digits = input.replaceAll("[^0-9]", "");
        if (digits.startsWith("00855")) {
            digits = digits.substring(5);
        } else if (digits.startsWith("855")) {
            digits = digits.substring(3);
        }
        if (!digits.startsWith("0")) {
            digits = "0" + digits;
        }
        if (digits.length() < 3) {
            return Optional.empty();
        }
        String prefix = digits.substring(0, 3);
        int subscriberLength = digits.length() - 3;
        boolean valid =
                (SIX_DIGIT.contains(prefix) && subscriberLength == 6)
                        || (SEVEN_DIGIT.contains(prefix) && subscriberLength == 7);
        return valid ? Optional.of("+855" + digits.substring(1)) : Optional.empty();
    }
}
